sys/net/if.c

```diff
@@ -3549,119 +3032,442 @@ ifioctl(struct socket *so, u_long cmd, caddr_t data, struct thread *td)
+			thunk.ifr.ifr_addr = ifr32->ifr_addr;
+			break;
+		case _IOC_NEWTYPE(SIOCSIFCAP, struct ifreq32):
+			thunk.ifr.ifr_reqcap = ifr32->ifr_reqcap;
+			break;
+		case _IOC_NEWTYPE(SIOCSIFMEDIA, struct ifreq32):
+			thunk.ifr.ifr_media = ifr32->ifr_media;
+			break;
+		case _IOC_NEWTYPE(SIOCSIFMTU, struct ifreq32):
+			thunk.ifr.ifr_mtu = ifr32->ifr_mtu;
+			break;
+		case _IOC_NEWTYPE(SIOCSIFPHYS, struct ifreq32):
+			thunk.ifr.ifr_phys = ifr32->ifr_phys;
+			break;
+		case _IOC_NEWTYPE(SIOCGI2C, struct ifreq32):
+			thunk.ifr.ifr_data = __USER_CAP(ifr32->ifr_data,
+			    sizeof(struct ifi2creq));
+			break;
+		case _IOC_NEWTYPE(SIOCSIFFIB, struct ifreq32):
+		case _IOC_NEWTYPE(SIOCSTUNFIB, struct ifreq32):
+			thunk.ifr.ifr_fib = ifr32->ifr_fib;
+			break;
+		case _IOC_NEWTYPE(SIOCSIFMETRIC, struct ifreq32):
+			thunk.ifr.ifr_metric = ifr32->ifr_metric;
+			break;
+		case _IOC_NEWTYPE(SIOCSIFVNET, struct ifreq32):
+		case _IOC_NEWTYPE(SIOCSIFRVNET, struct ifreq32):
+			thunk.ifr.ifr_jid = ifr32->ifr_jid;
+			break;
+		case _IOC_NEWTYPE(SIOCGIFGENERIC, struct ifreq32):
+		case _IOC_NEWTYPE(SIOCSIFGENERIC, struct ifreq32):
+		case _IOC_NEWTYPE(SIOCGVH, struct ifreq32):
+		case _IOC_NEWTYPE(SIOCIFCREATE2, struct ifreq32):
+			thunk.ifr.ifr_data =
+			    __USER_CAP_UNBOUND(ifr32->ifr_data);
+			break;
+		case _IOC_NEWTYPE(SIOCGETVLAN, struct ifreq32):
+		case _IOC_NEWTYPE(SIOCSETVLAN, struct ifreq32):
+			thunk.ifr.ifr_data = __USER_CAP(ifr32->ifr_data,
+			    sizeof(struct vlanreq));
+			break;
+		case _IOC_NEWTYPE(SIOCGIFDESCR, struct ifreq32):
+		case _IOC_NEWTYPE(SIOCSIFDESCR, struct ifreq32):
+			thunk.ifr.ifr_buffer.buffer =
+			    __USER_CAP(ifr32->ifr_buffer.buffer,
+				ifr32->ifr_buffer.length);
+			break;
+		case _IOC_NEWTYPE(SIOCSVH, struct ifreq32):
+			thunk.ifr.ifr_data = __USER_CAP(ifr32->ifr_data,
+			    sizeof(struct carpreq));
+			break;
+		case _IOC_NEWTYPE(SIOCSIFNAME, struct ifreq32):
+			thunk.ifr.ifr_data = __USER_CAP(ifr32->ifr_data,
+			    IFNAMSIZ);
+			break;
+		case _IOC_NEWTYPE(SIOCSLANPCP, struct ifreq32):
+			thunk.ifr.ifr_lan_pcp = ifr32->ifr_lan_pcp;
+			break;
+		case _IOC_NEWTYPE(SIOCG80211STATS, struct ifreq32):
+			thunk.ifr.ifr_data = __USER_CAP(ifr32->ifr_data,
+			    sizeof(struct ieee80211_stats));
+			break;
+		case _IOC_NEWTYPE(SIOCGETPFSYNC, struct ifreq32):
+		case _IOC_NEWTYPE(SIOCSETPFSYNC, struct ifreq32):
+			thunk.ifr.ifr_data = __USER_CAP(ifr32->ifr_data,
+			    sizeof(struct pfsyncreq));
+			break;
+		case _IOC_NEWTYPE(GIFGOPTS, struct ifreq32):
+		case _IOC_NEWTYPE(GIFSOPTS, struct ifreq32):
+			thunk.ifr.ifr_data = __USER_CAP(ifr32->ifr_data,
+			    sizeof(u_int));
+		case _IOC_NEWTYPE(IPSECGREQID, struct ifreq32):
+		case _IOC_NEWTYPE(IPSECSREQID, struct ifreq32):
+			thunk.ifr.ifr_data = __USER_CAP(ifr32->ifr_data,
+			    sizeof(uint32_t));
+			break;
+		}
+		saved_cmd = cmd;
+		saved_data = data;
+		data = (caddr_t)&thunk.ifr;
+		cmd = _IOC_NEWTYPE(cmd, struct ifreq);
```

Comment:
I keep thinking the translation should happen before the inner case statement, but can see it going either way.

Reply:
Yeah, I considered that.  I'm not sure which is cleaner.  I could do either way.